### PR TITLE
cql3: Add evaluate(expression) and use instead of term::bind()

### PR DESCRIPTION
**Pull request by cvybhu**
This PR adds the function:
```c++
constant evaluate(const expression&, const query_options&);
```
which evaluates the given expression to a constant value.
It binds all the bound values, calls functions, and reduces the whole expression to just raw bytes and `data_type`, just like `bind()` and `get()` did for `term`.

The code is often similar to the original `bind()` implementation in `lists.cc`, `sets.cc`, etc.

## Notes:
* For some reason in the original code, when a collection contains `unset_value`, then the whole collection is evaluated to `unset_value`. I'm not sure why this is the case, considering it's impossible to have `unset_value` inside a collection, because we forbid bind markers inside collections. For example here: https://github.com/scylladb/scylla/blob/cc8fc737614b537042bc6e6d5048628a8538677a/cql3/lists.cc#L134
This seems to have been introduced by Pekka Enberg in 50ec81ee671351254e60c9fa83d858bbedd0b9f0, but he has left the company.
I didn't change the behaviour, maybe there is a reason behind it, although maybe it would be better to just throw `invalid_request_exception`.
* There was a strange limitation on map key size, it seems incorrect: https://github.com/scylladb/scylla/blob/cc8fc737614b537042bc6e6d5048628a8538677a/cql3/maps.cc#L150, but I left it in.
* When evaluating a `user_type` value, the old code tolerated `unset_value` in a field, but it was later converted to NULL. This means that `unset_value` doesn't work inside a `user_type`, I didn't change it, will do in another PR.
* We can't fully get rid of `bind()` yet, because it's used in `prepare_term` to return a `terminal`. It will be removed in the next PR, where we finally get rid of `term`.

### CODE REVIEW

**Comment by nyh** [comment]
Replies to some of your notes:
1. I don't know if there is a syntax to bind only one a single collection value. Consider that like we have `UPDATE tab SET col[123] = 'value'` we could allow the syntax `UPDATE tab SET col[?] = 'value'`. Or like we have `UPDATE tab SET col = col + ['a']` we could have `UPDATE tab SET col = col + [?]`. Again, I didn't check if this is supported, but I don't see an inherent reason why it shouldn't be supported - so there is no reason to have internal code which assumes it can't. We **should** make sure that we have tests for this feature.
2. The limit on map key length is a real thing, and should be kept (I don't know if it needs to appear in this specific code - I'm just talking in general). Cassandra documents it as being limited to 65536 bytes. I don't know if we documented this anywhere. We **should** have tests for this as well. If they were cql-pytest tests, they could be run against both Cassandra and Scylla to compare. 
3. Again, for user_type, before deciding what is the correct and what is an erroneous change, please make sure you have a test for it - better a test that runs both for Cassandra and Scylla.

File: concrete_types.hh [hunk 1]
```diff
@@ -33,15 +33,39 @@
 
 struct empty_type_impl final : public abstract_type {
     empty_type_impl();
+
+    virtual bool contains_set_or_map() const {
+        return false;
+    }
+
+    virtual bool contains_collection() const {
```
Comment (nyh) [inline concrete_types.hh:41]: Just making sure: contains_colection() is almost the same as contains_set_or_map() but also checks for a list?

**Comment by avikivity** [comment]
> This PR adds the function:
> 
> ```c++
> constant evaluate(const expression&, const query_options&);
> ```

As Cato constantly says, this should not return `constant` but raw_value_view or a related type. The type is part of the expression but not part of the value it returns. But I understand why you need it - currently expressions don't know their own types, so you must carry it around in the values like Python or other dynamically types languages.

> 
> which evaluates the given expression to a constant value.
> It binds all the bound values, calls functions, and reduces the whole expression to just raw bytes and `data_type`, just like `bind()` and `get()` did for `term`.
> 
> The code is often similar to the original `bind()` implementation in `lists.cc`, `sets.cc`, etc.
> 
> ## Notes:
> * For some reason in the original code, when a collection contains `unset_value`, then the whole collection is evaluated to `unset_value`. I'm not sure why this is the case, considering it's impossible to have `unset_value` inside a collection, because we forbid bind markers inside collections. For example here: https://github.com/scylladb/scylla/blob/cc8fc737614b537042bc6e6d5048628a8538677a/cql3/lists.cc#L134
>   
>   This seems to have been introduced by Pekka Enberg in [50ec81e](https://github.com/scylladb/scylla/commit/50ec81ee671351254e60c9fa83d858bbedd0b9f0), but he has left the company.
>   I didn't change the behaviour, maybe there is a reason behind it, although maybe it would be better to just throw `invalid_request_exception`.

Not changing behavior is the correct approach in a long series. Changing behavior should be separate, in a single patch where the focus is on the behavior change.

Regarding what the behavior should be - I think part of the prepare phase should be to note which bind variables can be unset and which cannot, and reject improper unset values early (so we don't have to worry about them later). Similarly areas that can accept unset values should check for them early. There are just two: `UPDATE ... SET x = :unset` and `INSERT INTO tab (non_key_column, ...) VALUES (:unset, ...)`. They are illegal even there if they are not the top-level of an expression.

> * There was a strange limitation on map key size, it seems incorrect so I removed it: https://github.com/scylladb/scylla/blob/cc8fc737614b537042bc6e6d5048628a8538677a/cql3/maps.cc#L150

Better not to do it, and send a separate patch later. This way, if it turns out incorrect we can just revert it. If it's embedded in a large series it's harder to understand what part of the series caused the regression.

> * When evaluating a `user_type` value, the old code tolerated `unset_value` in a field, but it was later erroneously converted to NULL. This means that `unset_value` doesn't work inside a `user_type`, I changed it to throw an `invalid_request_exception`.

Again better not to change behavior (though I think the change is correct).

> * We can't fully get rid of `bind()` yet, because it's used in `prepare_term` to return a `terminal`. It will be removed in the next PR, where we finally get rid of `term`.

That's fine. I'm also anxiously anticipating a series that is able to evaluate() `column_value`.

File: cql3/expr/expression.cc [hunk 2]
```diff
@@ -1468,5 +1468,13 @@ utils::chunked_vector<std::vector<managed_bytes_opt>> get_list_of_tuples_element
 
     return tuples_list;
 }
+
+expression to_expression(const ::shared_ptr<term>& term_ptr) {
+    if (term_ptr.get() == nullptr) {
+        return constant::make_null();
```
Comment (avikivity) [inline cql3/expr/expression.cc:1938]: Why is a nullptr term equivalent to a CQL NULL? Isn't it an internal error?

File: cql3/lists.cc [hunk 3]
```diff
@@ -177,6 +181,10 @@ lists::marker::bind(const query_options& options) {
     }
 }
 
+expr::expression lists::marker::to_expression() {
+    throw std::runtime_error(fmt::format("to_expression not implemented! {}:{}", __FILE__, __LINE__));
```
Comment (avikivity) [inline cql3/lists.cc]: on_internal_error() is nicer, and so is std::source_location(), but I won't insist since this is all temporary code.

File: cql3/term.hh [hunk 4]
```diff
@@ -102,9 +102,7 @@ public:
         return out << t.to_string();
     }
 
-    expr::expression to_expression() {
-        throw std::runtime_error("unimplemented");
-    }
+    virtual expr::expression to_expression() = 0;
```
Comment (avikivity) [inline cql3/term.hh:105]: Perhaps it should be const.

Comment (avikivity) [reply cql3/term.hh:105]: Ah, can't, because term::get() isn't const.

File: cql3/expr/expression.hh [hunk 5]
```diff
@@ -208,6 +208,11 @@ struct bind_variable {
     // FIXME: infer shape from expression rather than from grammar
     shape_type shape;
     int32_t bind_index;
+
+    // Type of the bound value.
+    // Before preparing can be nullptr.
+    // After preparing always holds a valid type.
+    data_type value_type;
```
Comment (avikivity) [inline cql3/expr/expression.hh:230]: This can be merged with the patch that assigns it.

File: cql3/lists.cc [hunk 6]
```diff
@@ -182,7 +182,11 @@ lists::marker::bind(const query_options& options) {
 }
 
 expr::expression lists::marker::to_expression() {
-    throw std::runtime_error(fmt::format("to_expression not implemented! {}:{}", __FILE__, __LINE__));
+    return expr::bind_variable {
+        .shape = expr::bind_variable::shape_type::scalar,
```
Comment (avikivity) [inline cql3/lists.cc:197]: Is it really a scalar?

I expect not, but also that it doesn't matter, and the shape was used in deciding which marker subtype to create during prepare.

File: cql3/tuples.cc [hunk 7]
```diff
@@ -71,11 +71,19 @@ cql3::raw_value tuples::in_value::get(const query_options& options) {
 }
 
 expr::expression tuples::marker::to_expression() {
-    throw std::runtime_error(fmt::format("to_expression not implemented! {}:{}", __FILE__, __LINE__));
+    return expr::bind_variable {
+        .shape = expr::bind_variable::shape_type::tuple,
```
Comment (avikivity) [inline cql3/tuples.cc:85]: I see you did consider it. In any case I hope we can get rid of shape soon.

File: cql3/expr/expression.hh [hunk 8]
```diff
@@ -574,6 +574,15 @@ constant evaluate_IN_list(term&, const query_options&);
...
 
+// Takes a prepared expression and calculates its value.
+// Evaluates bound values, calls functions and returns just the bytes and type.
+constant evaluate(const expression& e, const query_options&);
+constant evaluate(const bind_variable&, const query_options&);
+constant evaluate(const tuple_constructor&, const query_options&);
+constant evaluate(const collection_constructor&, const query_options&);
+constant evaluate(const usertype_constructor&, const query_options&);
+constant evaluate(const function_call&, const query_options&);
+
```
Comment (avikivity) [inline cql3/expr/expression.hh:600]: Do these need to be in a header? The entry point does, but for the subtypes, they'll always be dispatched from the entrypoint, never called directly.

File: concrete_types.hh (hunk 1 above)
Comment (avikivity) [reply concrete_types.hh:41]: It's better to add this as a visitor, we're trying to simplify abstract_type.

Comment (avikivity) [reply concrete_types.hh:41]: (see types.cc for many visitors)

File: cql3/expr/expression.cc [hunk 9]
```diff
@@ -1304,6 +1304,10 @@ constant evaluate(term* term_ptr, const query_options& options) {
         return constant::make_null();
     }
 
+    if (dynamic_cast<abstract_marker*>(term_ptr) != nullptr) {
+        return evaluate(term_ptr->to_expression(), options);
+    }
```
Comment (avikivity) [inline cql3/expr/expression.cc]: I don't understand this. Why do we even have evaluate(term*)? Shouldn't we convert to expressions during prepare, then call evaluate(expression)?

This is going to be very slow, term->expression conversion is not supposed to be in a fast path.

File: cql3/expr/expression.cc [hunk 10]
```diff
@@ -1411,8 +1415,172 @@ constant evaluate(const expression& e, const query_options& options) {
     }, e);
 }
 
+// Checks whether a bound value of this type has to be reserialized
+static bool needs_to_be_reserialized(const abstract_type& type, const cql_serialization_format& sf) {
+    // If a value contains set or map, then this collection can be sent in the wrong order
```
Comment (avikivity) [inline cql3/expr/expression.cc:1415]: This can be in types.cc, the knowledge of serialization formats should be embedded there. 

Further, we can optimize it by computing it during type construction time and storing it in in a member.

File: cql3/expr/expression.cc [hunk 11]
```diff
@@ -1411,8 +1415,172 @@ constant evaluate(const expression& e, const query_options& options) {
...
+
+    return false;
+}
+
+// Serializes a map using the internal cql serialization format
+// Takes a range of pair<managed_bytes, managed_bytes>
+template <typename Range>
+requires requires (Range map_range) {
+    {*map_range.begin()} -> std::convertible_to<const std::pair<managed_bytes, managed_bytes>&>;
+}
```
Comment (avikivity) [inline cql3/expr/expression.cc:1435]: You can write:

```c__
template <std::ranges::range Range>
requires std::convertible_to<Range, const std::pair<...>>
```

Comment (avikivity) [reply cql3/expr/expression.cc:1435]: Don't we have such code already?

File: test/boost/user_types_test.cc [hunk 12]
```diff
@@ -580,7 +580,7 @@ SEASTAR_TEST_CASE(test_nonfrozen_user_types_prepared) {
         BOOST_REQUIRE_EXCEPTION(
             execute_prepared("insert into cf (a, b) values (?, ?)", {mk_int(4), mk_longer_tuple({4, "text4", int64_t(4), 5})}),
             exceptions::invalid_request_exception,
-            exception_predicate::message_equals("User Defined Type value contained too many fields (expected 3, got 4)"));
+            exception_predicate::message_contains("contained too many fields (expected 3, got 4)"));
```
Comment (avikivity) [inline test/boost/user_types_test.cc:583]: Why this change?

File: types.cc [hunk 13]
```diff
@@ -1512,6 +1512,18 @@ static void validate_aux(const tuple_type_impl& t, View v, cql_serialization_for
...
+    while (!v.empty()) {
+        read_tuple_element(v);
+        extra_elements += 1;
+    }
+
+    if (extra_elements > 0) {
+        // This function is called for both tuple and user_type, print the name too
+        throw marshal_exception(format("Value of type {} contained too many fields (expected {}, got {})",
+                                t.name(), t.size(), t.size() + extra_elements));
+    }
```
Comment (avikivity) [inline types.cc:1526]: Why this change?

File: cql3/expr/expression.cc [hunk 14]
```diff
@@ -1304,7 +1304,9 @@ constant evaluate(term* term_ptr, const query_options& options) {
         return constant::make_null();
     }
 
-    if (dynamic_cast<abstract_marker*>(term_ptr) != nullptr) {
+    if (dynamic_cast<abstract_marker*>(term_ptr) != nullptr
+    || dynamic_cast<tuples::value*>(term_ptr) != nullptr
+    || dynamic_cast<tuples::delayed_value*>(term_ptr) != nullptr) {
```
Comment (avikivity) [inline cql3/expr/expression.cc]: Bad indentation.

File: cql3/values.hh [hunk 15]
```diff
@@ -262,6 +262,18 @@ public:
...
     }
+    managed_bytes_opt to_managed_bytes_opt() && {
+        return std::visit(overloaded_functor{
+            [](bytes&& bytes_val) { return managed_bytes_opt(bytes_val); },
+            [](managed_bytes&& managed_bytes_val) { return managed_bytes_opt(std::move(managed_bytes_val)); },
+            [](null_value&&) -> managed_bytes_opt {
+                return std::nullopt;
+            },
+            [](unset_value&&) -> managed_bytes_opt {
+                return std::nullopt;
```
Comment (avikivity) [inline cql3/values.hh:273]: This is suspicious. Maybe we should throw here.

File: cql3/functions/functions.cc [hunk 16]
```diff
@@ -511,7 +511,8 @@ expr::expression function_call::to_expression() {
 
     return expr::function_call {
         .func = _fun,
-        .args = std::move(args)
+        .args = std::move(args),
```
Comment (avikivity) [inline cql3/functions/functions.cc:514]: Tip: always terminate lists with a trailing comma, as a courtesy to the next person to append to the list.

File: cql3/expr/expression.cc [hunk 17]
```diff
@@ -1310,29 +1310,7 @@ constant evaluate(term* term_ptr, const query_options& options) {
...
-    || dynamic_cast<lists::delayed_value*>(term_ptr) != nullptr 
-    || dynamic_cast<sets::value*>(term_ptr) != nullptr
-    || dynamic_cast<sets::delayed_value*>(term_ptr) != nullptr
-    || dynamic_cast<maps::value*>(term_ptr) != nullptr
-    || dynamic_cast<maps::delayed_value*>(term_ptr) != nullptr
-    || dynamic_cast<user_types::value*>(term_ptr) != nullptr
-    || dynamic_cast<user_types::delayed_value*>(term_ptr) != nullptr
-    || dynamic_cast<functions::function_call*>(term_ptr) != nullptr) {
-        return evaluate(term_ptr->to_expression(), options);
-    }
```
Comment (avikivity) [inline cql3/expr/expression.cc]: I don't see removals of the old code. Is it still reachable?

In general the best way to construct such series is to pair new code + removal of old code in the same patch. This way the reviewer can match old/new code, and can see we don't increase bloat.

**Comment by avikivity** [comment]
I'm worried about performance, since you call to_expression() in the fast path.

Please check with `build/release/test/perf/perf_simple_query --smp 1 -m 1G --operations-per-shard 1000000 --task-quota-ms 10`. The op rate will fluctuate, but insns/op and allocs/op should be stable.

**Comment by cvybhu** [comment]
v2:
* Removed all changes of behaviour, they will be done in separate PRs
* changed `conains_collection` to visitor
* Moved `needs_reserializing` to `abstract_type`
* other small fixes

File: cql3/term.hh [hunk 18]
```diff
@@ -100,6 +101,10 @@ public:
     friend std::ostream& operator<<(std::ostream& out, const term& t) {
         return out << t.to_string();
     }
+
+    expr::expression to_expression() {
```
[the remark this answers is not in the document]
Comment (cvybhu) [reply cql3/term.hh]: I wanted to show why reorganizing includes is needed - to introduce this function. I can remove it, but I feel that it makes reviewing the commits easier.

File: concrete_types.hh (hunk 1 above)
Comment (cvybhu) [reply concrete_types.hh]: > Just making sure: contains_colection() is almost the same as contains_set_or_map() but also checks for a list?

yes, added a blurb to make this more clear

> It's better to add this as a visitor, we're trying to simplify abstract_type.

:+1:, changed to visitor, there is much less code now

File: cql3/expr/expression.cc (hunk 2 above)
Comment (cvybhu) [reply cql3/expr/expression.cc:1932]: `nullptr` `terminal` means `null`.
For example in `constants::null_literal::null_value::bind`:
https://github.com/scylladb/scylla/blob/acf8da2bcef47168376e876e2fdc37fc26ebc82e/cql3/constants.hh#L86

Later it can be cast to term and then `to_expression` could be called on it.
In that case a nullptr `term` means a valid null value.

File: cql3/expr/expression.hh (hunk 5 above)
Comment (cvybhu) [reply cql3/expr/expression.hh:230]: done

File: cql3/lists.cc (hunk 6 above)
Comment (cvybhu) [reply cql3/lists.cc:197]: Tbh I'm not sure about it, but a `scalar` was the closest thing to a list value.

File: cql3/expr/expression.hh (hunk 8 above)
Comment (cvybhu) [reply cql3/expr/expression.hh:600]: I thought it could be useful. Maybe in the future there will be a situation when I have a `collection_constructor` and want to evaluate it (somewhere in prepare?). Without those functions I would have to create an `expression` and call evaluate on it, which is a lot of needless work.

File: cql3/expr/expression.cc (hunk 10 above)
Comment (cvybhu) [reply cql3/expr/expression.cc]: Do types ever change?
I think it's possible to add new field to a user type (?)

I was worried that this would be premature optimalization, the types aren't that big and there would be some new delicate bug prone logic introduced.

File: cql3/expr/expression.cc (hunk 11 above)
Comment (cvybhu) [reply cql3/expr/expression.cc:1418]: > You can write: `requires std::convertible_to<Range, const std::pair<...>>`

Range is in this case `std::map`, it can't be convertible to `std::pair`

I changed it to:
```c++
template <std::ranges::range Range>
requires requires (Range map_range) {
    {*std::ranges::begin(map_range)} -> std::convertible_to<const std::pair<managed_bytes, managed_bytes>&>;
}
```
I didn't do much concepts, so there might be some easier way that I'm not seeing

> Don't we have such code already?

The existing code does it the same way:
https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/maps.cc#L88

I looked around a bit but didn't find anything I could use for this

File: test/boost/user_types_test.cc (hunk 12 above)
Comment (cvybhu) [reply test/boost/user_types_test.cc:583]: The message changed because I moved validating `user_type` to `abstract_type::validate` (see the next comment).
It now contains `user_type::name()` which is very long and user type specific, so I changed it to `contains()`.

File: types.cc (hunk 13 above)
Comment (cvybhu) [reply types.cc:1526]: When evaluating a `bind_variable` we need to check that the serialized value represents the expected type.
This is done by calling `abstract_type::validate`.

For some reason `user_types::marker::bind` doesn't use `validate`.
[user_types::marker::bind](https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/user_types.cc#L141)  calls [user_types::value user_types::value::from_serialized](https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/user_types.cc#L57) which deserializes the elements and checks that there are no more than `size()` of them. 

That would mean that there is actually no validation for fields of user types value (???)

I wanted to have all validation done in `validate`, so I added checking number of fields to `abstract_type::validate` for `tuple_type_impl` (which also handles `user_type_impl` thanks to inheritance)

File: cql3/expr/expression.cc (hunk 14 above)
Comment (cvybhu) [reply cql3/expr/expression.cc]: It's temporary, but fixed.

File: cql3/values.hh (hunk 15 above)
Comment (cvybhu) [reply cql3/values.hh:273]: I wanted to have behaviour consistent with [`raw_value_view::to_managed_bytes_opt`](https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/values.hh#L158)

Making it throw would probably be more correct, but I'm worried that some existing code might break.

File: cql3/functions/functions.cc (hunk 16 above)
Comment (cvybhu) [reply cql3/functions/functions.cc:514]: :+1:, will do in the future

File: cql3/expr/expression.cc (hunk 17 above)
Comment (cvybhu) [reply cql3/expr/expression.cc]: Yes old `bind()` implementations are still reachable, because [`prepare_term` uses `bind()` to return a `terminal`](https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/expr/term_expr.cc#L410)
It can't be replaced with `evaluate` because `expr::constant` is not convertible to `terminal`.

Old code will be removed along with `term` in the next PR.

**Comment by cvybhu** [comment]
> I'm worried about performance, since you call to_expression() in the fast path.
>
> Please check with `build/release/test/perf/perf_simple_query --smp 1 -m 1G --operations-per-shard 1000000 --task-quota-ms 10.` The op rate will fluctuate, but insns/op and allocs/op should be stable.

I ran it on my laptop and the results are:
* current master (after revert, without PR #9244) (dd9d6c081ee48e670eda70e14a765442db62c0ec):
```
187120.44 tps ( 74.0 allocs/op,  12.0 tasks/op,   45267 insns/op)

median 187120.44 tps ( 74.0 allocs/op,  12.0 tasks/op,   45267 insns/op)
median absolute deviation: 0.00
maximum: 187120.44
minimum: 187120.44
```
* master with PR #9244 (15819e0304cbc488ddb4369e200f0a9f7e838530):
```
182394.19 tps ( 76.0 allocs/op,  12.1 tasks/op,   45751 insns/op)

median 182394.19 tps ( 76.0 allocs/op,  12.1 tasks/op,   45751 insns/op)
median absolute deviation: 0.00
maximum: 182394.19
minimum: 182394.19
```

* this branch:
```
174913.36 tps ( 75.0 allocs/op,  12.0 tasks/op,   45344 insns/op)

median 174913.36 tps ( 75.0 allocs/op,  12.0 tasks/op,   45344 insns/op)
median absolute deviation: 0.00
maximum: 174913.36
minimum: 174913.36
```

**Comment by cvybhu** [comment]
v3:
* Fixed indents
* Also changed `evaluate_IN_values` to use only `expression`

File: cql3/expr/expression.cc (hunk 2 above)
Comment (avikivity) [reply cql3/expr/expression.cc:1925]: Ok. I'll be happy when we can get rid of term.

File: cql3/expr/expression.cc (hunk 9 above)
Comment (avikivity) [reply cql3/expr/expression.cc]: Sure

Comment (avikivity) [reply cql3/expr/expression.cc]: Not so sure. We may branch 4.6 in between, leaving 4.6 with the regression.

File: cql3/expr/expression.cc (hunk 10 above)
Comment (avikivity) [reply cql3/expr/expression.cc]: It's possible to add new fields to user types, but they remain serialization compatible.

Serialization is not a property of the CQL front end, it's a property of the binary protocol between client and server and of the types themselves. It's really doesn't belong in expression.cc.

It's unlikely to ever change, since even if we add types, they won't exist in protocol v2.

File: cql3/expr/expression.cc (hunk 11 above)
Comment (avikivity) [reply cql3/expr/expression.cc:1389]: Map's value_type is `std::pair<const Key, Value>`. Note the const.

Comment (avikivity) [reply cql3/expr/expression.cc:1389]: And I should have written:

```c++
template <std::ranges::range Range>
requires std::convertible_to<std::ranges::range_value_t<Range>, std::pair<...>>
```

File: test/boost/user_types_test.cc (hunk 12 above)
Comment (avikivity) [reply test/boost/user_types_test.cc:583]: ok

File: types.cc (hunk 13 above)
Comment (avikivity) [reply types.cc:1526]: Please check if there is indeed no validation.

File: cql3/expr/expression.cc (hunk 17 above)
Comment (avikivity) [reply cql3/expr/expression.cc]: I see. Maybe it's better to merge the two series. It's better to have two large series than one very large series, but in this case we have a performance regression (or maybe not, we can measure) plus it leaves a lot of duplicate code.

I think we could have preserved performance by having expr::term_expr and expr_term : term, and converting between them during prepare. But let's not rewrite the whole series.

**Comment by avikivity** [comment]
Do you know where the extra allocation comes from?

The regression is small enough to allow into master.

File: types.cc (hunk 13 above)
Comment (cvybhu) [reply types.cc:1526]: yup, I made an example:
```rust
use scylla::{Session, SessionBuilder};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let session: Session = SessionBuilder::new()
        .known_node("127.0.0.1:9042")
        .build()
        .await?;

    session.query("CREATE KEYSPACE IF NOT EXISTS ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}", &[]).await?;
    session.query("CREATE TYPE IF NOT EXISTS ks.list_type (a frozen<list<int>>)", &[]).await?;
    session.query("CREATE TABLE IF NOT EXISTS ks.list_table (p int primary key, l list_type)", &[]).await?;

    let prepared = session.prepare("INSERT INTO ks.list_table (p, l) values (0, ?)").await?;

    // Insert tuple<int> instead of tuple<list<int>> as a value of type list_type, DB should refuse
    session.execute(&prepared, ((1234,),)).await?;

    Ok(())
}
```
This example finishes without errors and successfully inserts an `int32` in place of a `list<int>`
Then when I try to select using `cqlsh` I get:
```
cqlsh> select * from ks.list_table;
Traceback (most recent call last):
  File "/usr/local/bin/cqlsh", line 1042, in perform_simple_statement
    result = future.result()
  File "cassandra/cluster.py", line 4894, in cassandra.cluster.ResponseFuture.result
    raise self._final_exception
DriverException: Failed decoding result column "l" of type frozen<list_type>: Buffer slice out of bounds

```

At first I tried to make an example using python driver, but it turns out that python driver does type validation on its own and it refused to even send the query.

On this branch the example fails, although the error message is a bit ugly:
```
Error: DbError(Invalid, "Exception while binding value of type org.apache.cassandra.db.marshal.UserType(ks,6c6973745f74797065,61:org.apache.cassandra.db.marshal.FrozenType(org.apache.cassandra.db.marshal.ListType(org.apache.cassandra.db.marshal.Int32Type))): marshaling error: read_simple - not enough bytes (expected 4, got 0) Backtrace: 0x23c3b6e 0x23c4060 0x23c4368 0xcb4e5f 0xcb4ad4 0xcb4973 0xcb47d4 0x13d832a 0x13bd931 0x13bdb67 0x13c5906 0x11c08f9 0x11d46b1 0x11c02fb 0x11c045c 0x152152d 0x1297fc2 0x12983cd 0x1283226 0x1287df1 0x128b651 0x1282651 0x12846eb 0x1283773 0x128c257 0x128c7fc 0x1289804 0x224a115 0x228936f 0x228a577 0x22899ba 0x2241198 0x2240746 0xb94b65 /lib64/libc.so.6+0x27b74 0xb93a6d\n   --------\n   seastar::lambda_task<seastar::execution_stage::flush()::$_5>")
```

File: cql3/expr/expression.cc (hunk 11 above)
Comment (cvybhu) [reply cql3/expr/expression.cc]: That's nice, changed.

**Comment by cvybhu** [comment]
v4:
* nicer range syntax in `serialize_map`

**Comment by avikivity** [comment]
What about my question about allocation?

Also, doesn't #9244 have to be reinstated first?

**Comment by cvybhu** [comment]
> Also, doesn't #9244 have to be reinstated first?

Yes, I opened #9371

> What about my question about allocation?

Looking at it

**Comment by cvybhu** [comment]
> Do you know where the extra allocation comes from?

Allocations increased after the previous PR, it might because:
* Sometimes we need to create a new type instance as `shared_ptr`: https://github.com/cvybhu/scylla/blob/6bf6b03d12130c318d9748e03872b7a95c4b2712/cql3/restrictions/multi_column_restriction.hh#L322
* There might be cases where we needlessly serialize a value and then deserialize it back to access the elements. For example `lists::value` used to contain `vector<managed_bytes_opt>`, but now we serialize this vector.

I tried to get a more exact answer using some profiling tool, but didn't manage to do it.
I tried using gdb to print a backtrace each time `seastar::memory::allocate` is called, but this caused scylla to slow down so much that it broke.

There is some heap profiling in seastar, but when I tried using it there were errors, looks like it's old and abandoned.
https://github.com/scylladb/scylla/wiki/Seastar-heap-profiler
https://github.com/scylladb/scylla/blob/e1b22b6a4c56b4f1d0adf65d1a11db4bcb51fe7d/scylla-gdb.py#L1439

**Comment by cvybhu** [comment]
Rebased onto master

**Comment by cvybhu** [comment]
v5:
* optimized `contains_set_or_map` and `contains_collection`

**Comment by avikivity** [comment]
> > Do you know where the extra allocation comes from?
> 
> Allocations increased after the previous PR, it might because:
> 
> * Sometimes we need to create a new type instance as `shared_ptr`: https://github.com/cvybhu/scylla/blob/6bf6b03d12130c318d9748e03872b7a95c4b2712/cql3/restrictions/multi_column_restriction.hh#L322
> * There might be cases where we needlessly serialize a value and then deserialize it back to access the elements. For example `lists::value` used to contain `vector<managed_bytes_opt>`, but now we serialize this vector.

These are unlikely to be the cause, perf_simple_query doesn't use lists or multi_column_restriction.
